### PR TITLE
add login feature

### PR DESCRIPTION
**Pull request by edengo1996**

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -7,15 +7,25 @@
     app.run()
 
 
+@app.route('/test', methods=['POST'])
+def test_path2():
```
Comment (maorgo) [inline app.py:11]: If it's only test, let's remove it.

File: app.py [hunk 2]
```diff
@@ -7,15 +7,25 @@
...
+    return True
+
+
+@app.route('/test')
+def test_path():
```
Comment (maorgo) [inline app.py:16]: If it's only test, let's remove it.

File: app.py [hunk 3]
```diff
@@ -7,15 +7,25 @@
...
 @app.route('/')
 def hello_world():
     conn = sqlite3.connect('data_base.db')
     curs = conn.cursor()
     curs.execute('''CREATE TABLE IF NOT EXISTS users (
-                    Id INTEGER PRIMARY KEY AUTOINCREMENT,
-                    Username TEXT NOT NULL UNIQUE,
-                    Password TEXT NOT NULL
-                    )''')
+                       Id INTEGER PRIMARY KEY AUTOINCREMENT,
```
Comment (maorgo) [inline app.py:25]: Why did you add these spaces? seem unnecessary.

File: app.py [hunk 4]
```diff
@@ -7,15 +7,25 @@
...
+                       Username TEXT NOT NULL,
```
Comment (maorgo) [inline app.py:26]: Username should be UNIQUE. Why did you remove it?

File: app.py [hunk 5]
```diff
@@ -28,12 +38,43 @@ def sign_up_page():
...
+                 #(str_username, str_pass)).fetchall()
+
+    curs.execute('''SELECT * FROM users WHERE Username = ? AND Password = ?;''',
+                 (str_username, str_pass))
+    if len(curs.fetchall()) == 0:
+        return False
+    return True
+
+
+def process_sign_up():
```
Comment (maorgo) [inline app.py:77]: Separate it. You have views and logic. 
Logic code should be in a different class, views should only call for the logic processing.
Also, I want to see 2 separate stages: you first validate the arguments, then process it (save it to the DB).

File: app.py [hunk 6]
```diff
@@ -28,12 +38,43 @@ def sign_up_page():
 
 @app.route('/sign_up', methods=['POST'])
 def process_sign_up():
-    if sign_up():
-        return "process completed Eden", 201
-    return "failed to create a new user", 401
+    if process_sign_up():
+        return render_template('signup_completed.html'), 200
+    return render_template('signup_railed.html'), 400
```
Comment (maorgo) [inline app.py:43]: `signup_failed.html`, not `signup_railed`.
Also: why is that in a new file? it's the same, but just with an error message, right?

File: app.py [hunk 7]
```diff
@@ -28,12 +38,43 @@ def sign_up_page():
...
+
+
+@app.route('/login')
+def return_login_page():
+    return render_template('login_form.html')
+
 
+@app.route('/login', methods=['POST'])
+def process_login():
+    if login_verification():
```
Comment (maorgo) [inline app.py:53]: Same. Separate logic from views.

File: app.py [hunk 8]
```diff
@@ -28,12 +38,43 @@ def sign_up_page():
...
+        return render_template('login_completed.html'), 200
+    return render_template('login_failed.html'), 401
```
Comment (maorgo) [inline app.py:55]: Same comment. Login failed shouldn't be a new file but just an error inside the login page.
You also shouldn't have "login completed" file. 
Try to think of the minimum amount of HTML files you need here.

File: app.py [hunk 9]
```diff
@@ -28,12 +38,43 @@ def sign_up_page():
...
 
-def sign_up():
+
+def login_verification():
+    pass_text = request.form['pass']
+    username_text = request.form['name']
+    return existing_user(username_text, pass_text)
```
Comment (maorgo) [inline app.py:61]: You don't check if the username / password is None, you don't check if there are special characters, etc.

File: app.py [hunk 10]
```diff
@@ -28,12 +38,43 @@ def sign_up_page():
...
+
+
+def existing_user(str_username, str_pass):  # work here
```
Comment (maorgo) [inline app.py:64]: `does_user_exist` sound like a much better name for this function, no?
Also remove the comment here

File: app.py [hunk 11]
```diff
@@ -28,12 +38,43 @@ def sign_up_page():
...
+    conn = sqlite3.connect('data_base.db')
```
Comment (maorgo) [inline app.py:65]: Separate concerns: you should create one class that takes care of DB connections and use it where you need it, like here. Don't create this connection object here. What happens if you change the file name? how many other places need to be updated?

File: app.py [hunk 12]
```diff
@@ -28,12 +38,43 @@ def sign_up_page():
...
+    curs = conn.cursor()
+    # curs.execute('SELECT * FROM users WHERE Username = ? AND Password = ?',
```
Comment (maorgo) [inline app.py:67]: Why did you keep this comment?

File: app.py [hunk 13]
```diff
@@ -28,12 +38,43 @@ def sign_up_page():
...
+                 #(str_username, str_pass)).fetchall()
+
+    curs.execute('''SELECT * FROM users WHERE Username = ? AND Password = ?;''',
```
Comment (maorgo) [inline app.py:70]: Maybe wrap it inside a function in a different class? separate your SQL implementation from the views in flask. That way if you want to change something in the software, you only need to change it in one place.

File: app.py [hunk 14]
```diff
@@ -50,9 +91,19 @@ def sign_up():
 
 
 def check_valid_username(str_input):
-    conn = sqlite3.connect('database.db')
+    conn = sqlite3.connect('data_base.db')
```
Comment (maorgo) [inline app.py:94]: Same. separate SQL connection creation and views.

File: app.py [hunk 15]
```diff
@@ -50,9 +91,19 @@ def sign_up():
...
     curs = conn.cursor()
-    if curs.execute("SELECT ALL FROM users WHERE username = str_input ") == 0:
+    curs.execute('SELECT * FROM users WHERE Username = ?', (str_input,)).fetchall()
```
Comment (maorgo) [inline app.py:96]: Also here.

File: app.py [hunk 16]
```diff
@@ -50,9 +91,19 @@ def sign_up():
...
+    if len(curs.fetchall()) == 0:
         return True
     return False
 
+
```
Comment (maorgo) [inline app.py:101]: ????????

**Review by maorgo — CHANGES_REQUESTED** [review]
🔥 
A few comments, fix them and I'll go over it again.
Great job!

File: app.py (hunk 3 above)
Comment (edengo1996) [reply app.py:25]: if you are talking about lines 14 -18: than it's just spacing, since the query is fairly long.
is it not standard for us to space long queries?

File: app.py (hunk 4 above)
Comment (edengo1996) [reply app.py:26]: usernames ARE unique, as you can see we check for duplicates in 'check_valid_username' method

File: app.py (hunk 6 above)
Comment (edengo1996) [reply app.py:43]: updated 'signup_railed' to 'signup_failed'.
we use an html file so we can use a button to go back to our index page, instead of manually changing the URL.

File: app.py (hunk 7 above)
Comment (edengo1996) [reply app.py:53]: what do you mean by 'Seperate logic from views'?

File: app.py (hunk 10 above)
Comment (edengo1996) [reply app.py:64]: changed the name accordingly

File: app.py (hunk 9 above)
Comment (edengo1996) [reply app.py:61]: i don't understand. will call you soon for further investigation

File: app.py (hunk 11 above)
Comment (edengo1996) [reply app.py:65]: will call you for verification

File: app.py (hunk 12 above)
Comment (edengo1996) [reply app.py:67]: deleted

File: app.py (hunk 13 above)
Comment (edengo1996) [reply app.py:70]: will do

File: app.py (hunk 1 above)
Comment (edengo1996) [reply app.py:11]: removed

File: app.py (hunk 2 above)
Comment (edengo1996) [reply app.py:16]: removed

File: app.py (hunk 5 above)
Comment (edengo1996) [reply app.py:77]: got it

File: app.py (hunk 14 above)
Comment (edengo1996) [reply app.py:94]: got it

File: app.py (hunk 15 above)
Comment (edengo1996) [reply app.py:96]: got it

File: app.py (hunk 16 above)
Comment (edengo1996) [reply app.py:101]: got it

File: app.py (hunk 4 above)
Comment (maorgo) [reply app.py:26]: Did you fix it as we discussed?

**Review by maorgo — APPROVED** [review]
👍
